src/exa.c

```diff
@@ -47,27 +130,232 @@ static const uint8_t rop3[] = {
+    struct tegra_attrib_bo *old = scratch.attribs;
+    struct tegra_attrib_bo *new;
+    int err;
+
+    scratch.attribs_alloc_err = TRUE;
+
+    new = calloc(1, sizeof(*new));
+    if (!new)
+        return -1;
+
+    err = drm_tegra_bo_new(&new->bo, drm, 0, TEGRA_ATTRIB_BUFFER_SIZE);
+    if (err) {
+        free(new);
+        return err;
+    }
+
+    err = drm_tegra_bo_map(new->bo, (void**)&new->map);
+    if (err) {
+        drm_tegra_bo_unref(new->bo);
+        free(new);
+        return err;
+    }
+
+    scratch.attribs_alloc_err = FALSE;
+    scratch.attrib_itr = 0;
+    scratch.attribs = new;
+    new->next = old;
+
+    return 0;
+}
+
+static void TegraCompositeReleaseAttribBuffers(void)
+{
+    if (scratch.attribs) {
+        struct tegra_attrib_bo *attribs_bo = scratch.attribs;
+        struct tegra_attrib_bo *next;
+
+        while (attribs_bo) {
+            next = attribs_bo->next;
+            drm_tegra_bo_unref(attribs_bo->bo);
+            attribs_bo = next;
+        }
+
+        scratch.attribs = NULL;
+        scratch.attrib_itr = 0;
+    }
+}
+
+/*
+ * In general jobs submission overhead is higher for small pixmaps than
+ * overhead of handling them is SW.
+ */
+static Bool TegraEXASmallPixmap(DrawablePtr pDrawable)
+{
+    return (pDrawable->width *
+            pDrawable->height *
+            pDrawable->bitsPerPixel / 8 <= TEGRA_EXA_SMALL_PIXMAP_SIZE);
+}
+
+static void TegraEXAWaitFence(struct tegra_fence *fence)
+{
+    if (fence) {
+        if (tegra_stream_wait_fence(fence) && !fence->gr2d) {
+            /*
+             * XXX: A bit more optimal would be to release buffers
+             *      right after submitting the job, but then BO reservation
+             *      support by kernel driver is required. For now we will
+             *      release buffers when it is known to be safe, i.e. after
+             *      fencing which should happen often enough.
+             */
+            TegraCompositeReleaseAttribBuffers();
+        }
+    }
+}
+
+unsigned int TegraEXAPitch(unsigned int width, unsigned int height,
+                           unsigned int bpp)
+{
+    unsigned int alignment = 64;
+
+    /* GR3D texture sampler has specific alignment restrictions. */
+    if (width == 1 << LOG2_SIZE(width) &&
+        height == 1 << LOG2_SIZE(height))
+            alignment = 16;
+
+    return TEGRA_PITCH_ALIGN(width, bpp, alignment);
 }
 
 static int TegraEXAMarkSync(ScreenPtr pScreen)
 {
-    /* TODO: implement */
+    ScrnInfoPtr pScrn = xf86ScreenToScrn(pScreen);
+    TegraEXAPtr tegra = TegraPTR(pScrn)->exa;
 
-    return 0;
+    /* yes, we are converting pointer to integer, it is fine on ARM */
+    return (intptr_t) ((void *) tegra_stream_get_last_fence(&tegra->cmds));
 }
 
 static void TegraEXAWaitMarker(ScreenPtr pScreen, int marker)
 {
-    /* TODO: implement */
+    struct tegra_fence *fence = (void *) marker;
+
+    TegraEXAWaitFence(fence);
+
+    tegra_stream_put_fence(fence);
```

Comment:
Could we maybe get things like fencing introduced in a separate commit?

Reply:
Ok